### DISCUSSION TITLE
[QUESTION] How does one mock zip.js in Jest tests?

### DISCUSSION THREAD

**Discussion by remarkablemark** [gildas-lormeau/zip.js#424] [Q&A]
I'm using `zip.js` in my frontend project and I'm trying to write Jest tests but they're failing with:

```
    ReferenceError: TransformStream is not defined

    > 1 | import { BlobReader, BlobWriter, ZipWriter } from '@zip.js/zip.js';
        | ^
```

I'm using `jsdom` for my test environment.

Is the best way to resolve this is to [mock the node module](https://jestjs.io/docs/manual-mocks#mocking-node-modules)? I have seen other projects provide a [mock file for Jest](https://github.com/stripe/stripe-react-native#testing).

**Comment by gildas-lormeau** [comment]
If the web steam API is not available, then you should able to circumvent this issue by importing this module before running the tests: https://www.npmjs.com/package/web-streams-polyfill

**Reply by janesenaj42** [reply]
Apparently `TransformStream` is [available in Node 18 ](https://nodesource.com/blog/NodeV18-Available-in-Distributions/). But for anyone on Angular+Jest on older node versions, this worked for me in Node 14.15.5

````
// setup-jest.ts

import 'jest-preset-angular/setup-jest'
import { TransformStream } from 'web-streams-polyfill/ponyfill';

Object.defineProperties(globalThis, {
   TransformStream: {
      value: TransformStream
   }
}
```

**Comment by remarkablemark** [comment]
Thanks! I ended up creating a module mock for the package:

```sh
mkdir -p __mocks__/@zip.js/
echo "module.exports = {};" > __mocks__/@zip.js/zip.js.js
```
